### PR TITLE
Unify our Image types

### PR DESCRIPTION
**Pull request by abadams**
This PR unifies our various image types. Previously we had `Halide::Tools::Image<T>`, `Halide::Image<T>`, and `Halide::Buffer`. There is now only `Halide::Buffer<T>`, which lives in runtime/HalideBuffer.h. The class is basically a copy-paste of `Halide::Tools::Image<T>`, with a few warts fixed. Buffers of unknown type are represented as `Halide::Buffer<>` (T defaults to void). Casting from `Halide::Buffer<>` to `Halide::Buffer<T>` does a runtime type check.

`Halide::Buffer<T>` wraps a buffer_t. It optionally owns the data, as with our existing Image classes. Device allocations are considered owned iff the host allocation is. 

There is also Halide::Internal::BufferPtr, which is a named shared pointer to a `Halide::Buffer<T>`. This type is not exposed to users.

`Halide::Image<T>` still exists in this PR as an alias for `Halide::Buffer<T>`. This was done to minimize the number of changed files. I'll delete `Halide::Image<T>` in a follow-up PR.

`Halide::Buffer<T>` actually has a second template argument which gives the maximum number of dimensions supported by the class. This is currently limited to 4 for obvious reasons. When it wraps halide_buffer_t this hard limit will go away, and instead it will indicate the point at which the shape array spills from living within the class to living in a separate heap allocation. In that glorious future, we'll just have `Halide::Buffer<T>` as our C++ buffer type, and halide_buffer_t as the underlying C type it wraps.

Note that the interface for querying the shape of a `Halide::Buffer<T>` looks like buf.dim(2).stride(), which mirrors how halide_buffer_t will act (buf.dim[2].stride). To make ImageParam match this interface too, I pulled the corresponding changes to it in new_buffer_t into this branch. buf.stride(2) and similar still work, but will be deprecated at some point in the future.

One thing we lost was the constraint that `Halide::Image<T>` always exists on the host. Now if you say func.realize(800, 600) it creates a new Image and does copy-back for you before returning it if necessary, and if you say func.realize(existing_image) it doesn't do a copy back, so there are a few new calls to copy_to_host in the tests.

Sorry for the amount of churn in the python bindings, there are a bunch of drive-by fixes in there. As another follow-up commit I'll just clang-format them.

### CODE REVIEW

**Comment by abadams** [comment]
Windows build being tested here: 

http://buildbot.halide-lang.org:8010/builders/win-64-trunk/builds/1055

**Comment by abadams** [comment]
Take three on the windows build:

http://buildbot.halide-lang.org:8010/builders/win-64-trunk/builds/1061

File: src/Buffer.h [hunk 1]
```diff
@@ -2,144 +2,107 @@
 #define HALIDE_BUFFER_H
 
 /** \file
- * Defines Buffer - A c++ wrapper around a buffer_t.
+ * Defines BufferPtr - A c++ wrapper around a buffer_t.
```
Comment (steven-johnson) [inline src/Buffer.h]: The filename is now out of convention -- should we change it to BufferPtr.h (in a following PR)?

Comment (abadams) [inline src/Buffer.h]: Done, because it might actually make the diff easier to read. BufferPtr is only superficially similar to Buffer

**Comment by steven-johnson** [comment]
LGTM, very nice. (Disclaimer: I skipped most of the python-binding changes.)

File: src/AddImageChecks.cpp [hunk 2]
```diff
@@ -251,7 +251,7 @@ Stmt add_image_checks(Stmt s,
             Expr actual_extent = Variable::make(Int(32), actual_extent_name, image, param, rdom);
             Expr actual_stride = Variable::make(Int(32), actual_stride_name, image, param, rdom);
             if (!touched[j].is_bounded()) {
-                user_error << "Buffer " << name
+                user_error << "BufferPtr " << name
```
Comment (dsharletg) [inline src/AddImageChecks.cpp]: I think it should remain `Buffer` in the error message. It's the buffer being accessed, not the pointer to the buffer (I'm guessing this was a find/replace artifact...).

File: src/BufferPtr.h [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+#include "runtime/HalideBuffer.h"
+#include "Expr.h"
+#include "Util.h"
+#include "Argument.h"
+
+namespace Halide {
+namespace Internal {
+struct BufferContents;
+
+/** A named reference-counted handle on an Image of unknown type and dimensionality */
```
Comment (dsharletg) [inline src/BufferPtr.h]: This comment seems stale (and also doesn't match the file comment).

File: src/runtime/HalideBuffer.h [hunk 4]
```diff
@@ -0,0 +1,1443 @@
...
+    /** Get the Halide type of T. Callers should not use the result if
+     * T is void. */
+    static halide_type_t static_halide_type() {
+        return halide_type_of<not_void_T>();
+    }
+
+    /** Increment the reference count of any owned allocation */
+    void incref() {
+        if (alloc) {
+            alloc->ref_count++;
```
Comment (dsharletg) [inline src/runtime/HalideBuffer.h]: Should these be atomic?

**Comment by dsharletg** [comment]
LGTM aside from a few minor issues

File: src/AddImageChecks.cpp (hunk 2 above)
Comment (abadams) [inline src/AddImageChecks.cpp]: Oops, yeah overzealous search/re[place

File: src/runtime/HalideBuffer.h (hunk 4 above)
Comment (abadams) [inline src/runtime/HalideBuffer.h]: They are. ref_count is a `std::atomic<int>`

**Pull request merged**